src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp

```diff
@@ -156,7 +156,29 @@ bool ShootOrPassPlayFSM::passFound(const Update& event)
 
 bool ShootOrPassPlayFSM::shouldAbortPass(const Update& event)
 {
-    // TODO (#2384): implement this
+    // lost possession
+    if (event.common.world.getTeamWithPossession() != TeamSide::FRIENDLY)
+    {
+        return true;
+    }
+
+    // distance between robot and ball is too far,
+    // i.e. team might still have possession, but kicker/passer doesn't have control over
+    // ball
+    if ((event.common.world.ball().position() -
+         best_pass_and_score_so_far.pass.passerPoint())
+            .length() >
+        this->ai_config->getShootOrPassPlayConfig()->getMinDistanceToPass()->value())
+    {
+        return true;
+    }
+
+    // ball has already been kicked
+    if (tookShot(event))
+    {
+        return true;
+    }
```

Comment:
wouldn't this mean that as soon as the ball is kicked, the pass should be aborted?